src/Service/ProcessUtils.cpp

```diff
@@ -0,0 +1,71 @@
+// Copyright (c) 2021 The Orbit Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#include "ProcessUtils.h"
+
+#include "OrbitBase/Logging.h"
+#include "ServiceUtils.h"
+
+namespace orbit_service {
+
+ErrorMessageOr<std::vector<orbit_grpc_protos::ModuleInfo>> ReadModulesFromProcMaps(int32_t pid) {
+  const auto proc_maps_data = orbit_elf_utils::ReadProcMapsFile(pid);
+  if (proc_maps_data.has_error()) return proc_maps_data.error();
+
+  const auto map_entries = orbit_elf_utils::ParseMaps(proc_maps_data.value());
+
+  std::vector<orbit_grpc_protos::ModuleInfo> result;
+  result.reserve(map_entries.size());
+
+  for (const orbit_elf_utils::MapEntry& entry : map_entries) {
+    if (!entry.is_executable) continue;
+
+    ErrorMessageOr<orbit_grpc_protos::ModuleInfo> module_info_or_error = ErrorMessage{};
+
+    if (entry.inode != 0) {
+      module_info_or_error = CreateModuleFromFile(entry);
+    } else {
+      module_info_or_error = CreateModuleFromProcessMemory(pid, entry);
+    }
+
+    if (module_info_or_error.has_error()) {
+      ERROR("Unable to create module: %s", module_info_or_error.error().message());
+      continue;
+    }
+
+    result.emplace_back(std::move(module_info_or_error.value()));
+  }
+
+  return result;
+}
+
+ErrorMessageOr<orbit_grpc_protos::ModuleInfo> CreateModuleFromProcessMemory(
+    int32_t pid, const orbit_elf_utils::MapEntry& map_entry) {
+  if (map_entry.end_address <= map_entry.start_address) {
+    return ErrorMessage{
+        absl::StrFormat("Invalid address range for module \"%s\".", map_entry.module_path)};
+  }
+
+  const auto size = map_entry.end_address - map_entry.start_address;
+
+  // 8 MiB - That's also the limit that we support in the `ReadProcessMemory` gRPC endpoint.
+  // It's more than enough for reading [vdso] which is typically around 8 KiB.
+  if (size > 8 * 1024 * 1024) {
```

Comment:
Passing large files over grpc channel could run into DEADLINE_EXCEEDED errors. I wonder if we still want to do that via dump to file and then scp like we do for other modules. scp does not have a deadline and we already have some progress indicators ready to use there.

Reply:
So you're saying the current limit for the  `GetProcessMemory` rpc call is already too high?

For my use-case I only need 8 KiB. Is that also going to be a problem? Supporting other potentially bigger virtual modules is something that we basically got for free, that's why I added that as well. But I could also remove that.

I feel that writing that to a file while the only use-case we know of won't rellay benefit from that, is a bit overkill.

Reply:
I remember we were having deadline problems when reading symbols and sending them over the network, I think it could repeat itself, not for vdso but if user desided to load something bigger (and dynamically allocated).

Reply:
I made some tests. And indeed 8 MiB are too large. There seems to be a limit in gRPC around 4 MiB.

So new I changed the limit to 3 MiB and made some tests by artificially hacking in a bigger size for vdso. Works fine on both of my machines.

So now the user will not see any modules larger than 3 MiB. I could also change the behaviour that modules are shown no matter which size, but we show an error message when the user tries to load symbols for them.

WDYT?

Reply:
Isn't 4 MB the default limit for gRPC messages?

Reply:
I remember that when we were sending symbols over gRPC we had to increase that limit somewhere.

Reply:
> Isn't 4 MB the default limit for gRPC messages?

Yes, it seems so. I took the 8 MiB because the `GetProcessMemory` callback limits its result to 8 MiB, which would obviously not work and should be fixed as well.